checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java

```diff
@@ -1,176 +1,777 @@
+        continue;
+      }
 
-    } else {
-      if (!(cur instanceof SingleSuccessorBlock)) {
-        throw new BugInCF("BlockImpl is neither a conditional block nor a SingleSuccessorBlock");
+      VariableElement paramElt = TreeUtils.elementFromDeclaration(currentMethodParamTree);
+      if (!nodeAndElementResourceAliased(obligations, arg, paramElt)) {
+        continue;
       }
 
-      Block b = ((SingleSuccessorBlock) cur).getSuccessor();
-      if (b != null) {
-        successorBlock.add(b);
+      List<String> mustCallValues = resourceLeakAtf.getMustCallValues(paramElt);
+      // TODO: generalize this method to MustCall annotations with more than one element.
+      assert mustCallValues.size() <= 1 : "TODO: Handle larger must-call values sets";
+      Set<ResourceAlias> nodeAliases = getResourceAliasOfNode(obligations, arg);
+      for (ResourceAlias resourceAlias : nodeAliases) {
+        AnnotationMirror cmAnno = getCalledMethodsAnno(invocation, resourceAlias.reference);
+        if (mcca.calledMethodsSatisfyMustCall(mustCallValues, cmAnno)) {
+          addOwningToParam(i + 1);
+          break outerLoop;
+        }
       }
     }
-    return successorBlock;
+  }
+
+  /**
+   * Returns the set of resource aliases associated with the given node, by looking up the
+   * corresponding obligation in the given set of obligations.
+   *
+   * @param obligations the set of obligations to search in
+   * @param node the node whose resource aliases are to be returned
+   * @return the resource aliases associated with the given node, or an empty set if the node has
+   *     none
+   */
+  private Set<ResourceAlias> getResourceAliasOfNode(Set<Obligation> obligations, Node node) {
+    Node tempVar = mcca.getTempVarOrNode(node);
+    if (!(tempVar instanceof LocalVariableNode)) {
+      return Collections.emptySet();
+    }
+
+    Obligation argumentObligation =
+        MustCallConsistencyAnalyzer.getObligationForVar(obligations, (LocalVariableNode) tempVar);
+    if (argumentObligation == null) {
+      return Collections.emptySet();
+    }
+    return argumentObligation.resourceAliases;
+  }
+
+  /**
+   * Infers @Owning annotations for formal parameters of the enclosing method or fields of the
+   * enclosing class, as follows:
+   *
+   * <ul>
+   *   <li>If a formal parameter is passed as an owning parameter, add an @Owning annotation to that
+   *       formal parameter (see {@link #inferOwningParamsViaOwnershipTransfer}).
+   *   <li>It calls {@link #computeOwningForArgsOrReceiverOfCall} to compute @Owning annotations for
+   *       the receiver or arguments of a call by analyzing the called-methods set after the call.
+   * </ul>
+   *
+   * @param obligations the set of obligations to search in
+   * @param invocation the method or constructor invocation
+   */
+  private void computeOwningFromInvocation(Set<Obligation> obligations, Node invocation) {
+    if (methodElt == null) {
```

Comment:
A brief comment to explain this (seemingly-unrelated) check here would be helpful. `methodElt` is a field of the class and might be unrelated to `invocation`, right? So why does it need to be checked here?

Reply:
It's not actually dereferenced in the body of this method...

Reply:
It is used in `addOwningToParam`. I can't recall if the null value here has caused any NPE in real benchmarks. Even if it does, I agree that here is not the best place to check this. I will remove it from here.